graph-gateway/src/network/service.rs

```diff
@@ -0,0 +1,383 @@
+            indexers_min_scalar_tap_version: self.indexer_min_scalar_tap_version,
+            indexers_min_graph_node_version: self.indexer_min_graph_node_version,
+            indexers_addr_blocklist: self.indexer_addr_blocklist,
+            indexers_host_resolver: Mutex::new(self.indexer_host_resolver),
+            indexers_host_blocklist: self.indexer_host_blocklist,
+            indexers_pois_blocklist: self
+                .indexer_pois_blocklist
+                .map(|(bl, res)| (bl, Mutex::new(res))),
+            indexers_indexing_status_resolver: self.indexer_indexing_status_resolver,
+            indexers_cost_model_resolver: (
+                self.indexer_cost_model_resolver,
+                Mutex::new(self.indexer_cost_model_compiler),
+            ),
+        };
+
+        NetworkServicePending {
+            subgraph_client: self.subgraph_client,
+            internal_state,
+            update_interval: self.update_interval,
+        }
+    }
+}
+
+/// The [`NetworkService`] pending instance.
+///
+/// This struct represents the [`NetworkService`] instance that is pending spawning. To spawn the
+/// [`NetworkService`] instance, call the [`NetworkServicePending::spawn`] method.
+pub struct NetworkServicePending {
+    update_interval: Duration,
+    subgraph_client: SubgraphClient,
+    internal_state: InternalState,
+}
+
+impl NetworkServicePending {
+    /// Spawns the [`NetworkService`] instance's background task and returns the service
+    /// instance.
+    pub fn spawn(self) -> NetworkService {
+        let network = spawn_updater_task(
+            self.subgraph_client,
+            self.internal_state,
+            self.update_interval,
+        );
+
+        NetworkService { network }
+    }
+}
+
+/// Spawn a background task to fetch the network topology information from the graph network
+/// subgraph at regular intervals
+fn spawn_updater_task(
+    subgraph_client: SubgraphClient,
+    state: InternalState,
+    update_interval: Duration,
+) -> Eventual<Ptr<GraphNetwork>> {
+    let (mut eventual_writer, eventual) = Eventual::new();
+
+    tokio::spawn(async move {
+        let subgraph_client = Mutex::new(subgraph_client);
+        loop {
+            // Fetch the network topology information every `update_interval` duration
+            // If the fetch fails or takes too long, log a warning and skip the update
+            tokio::select! { biased;
+                update = fetch_update(&subgraph_client, &state) => {
+                    match update {
+                        Ok(network) => {
+                            eventual_writer.write(Ptr::new(network));
+                        }
+                        // If the fetch fails, log a warning and skip the update
+                        Err(err) => {
+                            tracing::warn!("network update fetch failed");
+                            tracing::debug!("network update fetch failed: {err}");
```

Comment:
Let's not duplicate logs like this. If we are ever looking for this log message, we probably want the full detail.
```suggestion
                            tracing::warn!("network update fetch failed: {err}");
```

Reply:
Also, I generally prefer the pattern of labeling the error instead of prefixing the message, like
```rust
Err(network_update_err) => tracing::warn!(%network_update_err),
```
I find it easier to filter logs this way, since it uses some structure (and it's nice to be able to scan for things like `_err`).